openslides_backend/action/motion_category/number_motions.py

```diff
@@ -0,0 +1,225 @@
+                        lead_motions_counter[lead] += 1
+                    else:
+                        number_value_map[motion_id] = 1
+                        lead_motions_counter[lead] = 2
+                else:
+                    number_value_map[motion_id] = main_counter
+                    main_counter += 1
+
+            for motion_id in affected_motions:
+                number = self.get_number(motion_id, number_value_map)
+                if number in existing_numbers:
+                    raise ActionException(
+                        f'Numbering aborted because the motion identifier "{number}" already exists.'
+                    )
+
+                yield {
+                    "id": motion_id,
+                    "number": number,
+                }
+
+    def init_memory(self, main_category_id: int) -> None:
+        """Preload all categories with needed fields, all motions with needed fields
+        and meeting with needed fields."""
+        category = self.datastore.get(
+            FullQualifiedId(Collection("motion_category"), main_category_id),
+            ["meeting_id"],
+        )
+
+        if category.get("meeting_id"):
+            meeting = self.datastore.get(
+                FullQualifiedId(Collection("meeting"), category["meeting_id"]),
+                ["motion_ids", "motion_category_ids"],
+            )
+        else:
+            meeting = {}
+
+        gmr_categories = GetManyRequest(
+            Collection("motion_category"),
+            meeting.get("motion_category_ids", []),
+            ["prefix", "parent_id", "child_ids", "weight", "motion_ids"],
+        )
+        gmr_motions = GetManyRequest(
+            Collection("motion"),
+            meeting.get("motion_ids", []),
+            [
+                "lead_motion_id",
+                "meeting_id",
+                "category_weight",
+                "category_id",
+                "number",
+            ],
+        )
+        gmr_meetings = GetManyRequest(
+            Collection("meeting"),
+            [category.get("meeting_id", 1)],
+            [
+                "motions_number_with_blank",
+                "motions_number_min_digits",
+                "motions_amendments_prefix",
+            ],
+        )
+        result = self.datastore.get_many([gmr_categories, gmr_motions, gmr_meetings])
+        self.mem_categories = result.get(Collection("motion_category"), {})
+        self.mem_motions = result.get(Collection("motion"), {})
+        self.mem_meetings = result.get(Collection("meeting"), {})
+
+    def get_prefix(self, category_id: int) -> str:
+        """Get the prefix of a category. If none, get the prefix of the parent if exists."""
+        category = self.mem_categories.get(category_id, {})
+        if category.get("prefix"):
+            return category["prefix"]
+        elif not category.get("parent_id"):
+            return ""
+        else:
+            return self.get_prefix(category["parent_id"])
```

Comment:
This is wrong: The prefix is only calculated within the affected categories. If you have two categories like this: `{id: 1, prefix: "A", name: "A"} -> {id: 2, name: "no prefix"}` and you call `number_motions` with id 2, no motion get a category prefix because category A is not affected. Please also add a test for this.

Reply:
Add a test and a check if the category is the main category.